### PR TITLE
[TP][DTensor Op] Enable Embedding op for DTensor

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96702

We enabled col-wise embedding for TP users.

### CODE REVIEW

File: torch/distributed/_tensor/ops/embedding_ops.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import torch
+
+from torch.distributed._tensor.api import DTensorSpec, Replicate, Shard
+from torch.distributed._tensor.op_schema import OpSchema, OutputSharding
+from torch.distributed._tensor.ops.utils import register_prop_rule
+
+aten = torch.ops.aten
+
+
+@register_prop_rule(aten.embedding.default)
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/embedding_ops.py:13]: I think we should also enable the backward of embedding call, maybe we can do it in a follow up PR.

File: torch/distributed/_tensor/ops/embedding_ops.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        raise NotImplementedError(
+            "DTensor does not support row-wise sharded embedding operation yet!"
+        )
+
+    if all(placement.is_replicate() for placement in inp_spec.placements):
+        return OutputSharding(
+            output_spec=DTensorSpec(
+                mesh=inp_spec.mesh,
+                placements=[Replicate()] * (len(inp_spec.placements) - 1) + [Shard(-1)],  # type: ignore[list-item]
+                tensor_meta=inp_spec.tensor_meta,
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/embedding_ops.py:26]: nit: for `output_sharding`, we don't need to set up the tensor meta field (can just leave it as None), as it would be overwritten by the sharding propagator itself.

File: torch/distributed/_tensor/ops/embedding_ops.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    if any(placement.is_shard(0) for placement in weight_spec.placements):
+        raise NotImplementedError(
+            "DTensor does not support row-wise sharded embedding operation yet!"
+        )
+
+    if all(placement.is_replicate() for placement in inp_spec.placements):
+        return OutputSharding(
+            output_spec=DTensorSpec(
+                mesh=inp_spec.mesh,
+                placements=[Replicate()] * (len(inp_spec.placements) - 1) + [Shard(-1)],  # type: ignore[list-item]
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/embedding_ops.py:25]: I think the output sharding of it would need to depend on the `weight_spec` but not the `in_spec`? It should basically the same as `weight_spec` placements I think (the output sharding layout should be the same as how the embedding is being sharded).

File: torch/distributed/_tensor/placement_types.py [hunk 4]
```diff
@@ -72,7 +72,8 @@ def _split_tensor(
 
     def _pad_tensor(self, tensor: torch.Tensor) -> torch.Tensor:
         # pad tensor by 1 on the shard dim
-        pad = [0, 0] * (tensor.ndim - self.dim)
+        pad_num = -self.dim if self.dim < 0 else tensor.ndim - self.dim
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:75]: So this is dealing with negative dims? My feeling is that we might want to keep Shard placement `dim` be always positive (unless specific reason) in our internal op implementation, if it's negative from user, we should call `normalize_dim` before constructing `Shard` placement, this is so that we deal with `negative_dim` in a uniform manner, see https://github.com/pytorch/pytorch/blob/60e503d4685f7cb4e7b7af5b11548c1d47d299e0/torch/distributed/_tensor/ops/utils.py#L62

**Review by wanchaol — COMMENTED** [review]
Nice work! Mostly looks good to me, have some suggestion on the output sharding and negative dimension handling

**Review by wanchaol — APPROVED** [review]
lgtm, my only comment is that let's call `normalize_dim` in `_pad_tensor` to unify the negative dim handling, thanks!

File: torch/distributed/_tensor/ops/embedding_ops.py (hunk 3 above)
Comment (wanchaol) [inline torch/distributed/_tensor/ops/embedding_ops.py:25]: I think a correct way to do this:
```
weight_dim_map = weight_spec.dim_map
output_dim_map = inp_spec.dim_map
output_dim_map.append(weight_dim_map[1])

output_spec = DTensorSpec.from_dim_map(mesh, output_dim_map, [])
```
